### PR TITLE
python3Packages.torchaudio: fix the cuda build

### PR DESCRIPTION
**Pull request by SomeoneSerge**
## Description of changes

Fixes https://hercules-ci.com/accounts/github/SomeoneSerge/derivations/%2Fnix%2Fstore%2Fxk9mwdpfq0pf8lr7vpdbwa5wh7n61jbk-python3.10-torchaudio-2.0.2.drv/log?via-job=34212fa1-acc7-4d65-ac40-a96f256e4911

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/python-modules/torch/default.nix [hunk 1]
```diff
@@ -436,11 +436,7 @@ in buildPythonPackage rec {
     blasProvider = blas.provider;
     # To help debug when a package is broken due to CUDA support
     inherit brokenConditions;
-  } // lib.optionalAttrs cudaSupport {
-    # NOTE: supportedCudaCapabilities isn't computed unless cudaSupport is true, so we can't use
-    #   it in the passthru set above because a downstream package might try to access it even
-    #   when cudaSupport is false. Better to have it missing than null or an empty list by default.
-    cudaCapabilities = supportedCudaCapabilities;
+    cudaCapabilities = if cudaSupport then supportedCudaCapabilities else [ ];
```
Comment (SomeoneSerge) [inline pkgs/development/python-modules/torch/default.nix:439]: @ConnorBaker objections?

File: pkgs/development/python-modules/torchaudio/default.nix [hunk 2]
```diff
@@ -27,17 +27,30 @@ buildPythonPackage rec {
...
     pkg-config
     ninja
   ] ++ lib.optionals cudaSupport [
-    cudaPackages.cudatoolkit
+    cudaPackages.cuda_nvcc
   ];
   buildInputs = [
     pybind11
   ] ++ lib.optionals cudaSupport [
-    cudaPackages.cudnn
```
Comment (SomeoneSerge) [inline pkgs/development/python-modules/torchaudio/default.nix:40]: torchaudio doesn't actually use cudnn (cf. e.g. https://github.com/pytorch/audio/issues/3472#issuecomment-1632634528)

File: pkgs/development/python-modules/torch/default.nix (hunk 1 above)
Comment (ConnorBaker) [reply pkgs/development/python-modules/torch/default.nix:439]: Eh, no objections here -- in retrospect, perhaps it *is* better to have an empty list than for it to just be missing.

We don't have a common API or way of exposing which capabilities a derivation is built with. Maybe something to look into for the future?

**Review by ConnorBaker — APPROVED** [review]
Evaluated and built fine on `x86_64-linux`. Thank you @SomeoneSerge!

**Comment by junjihashimoto** [comment]
@ConnorBaker Thank you for confirming @ConnorBaker !

**Comment by SomeoneSerge** [comment]
Aw, I didn't squash 🙃

**Pull request merged**